src/main/java/eu/europeana/metis/sandbox/service/workflow/HarvestServiceOaiPmhImpl.java

```diff
@@ -0,0 +1,69 @@
+package eu.europeana.metis.sandbox.service.workflow;
+
+import eu.europeana.metis.harvesting.HarvesterException;
+import eu.europeana.metis.harvesting.ReportingIteration;
+import eu.europeana.metis.harvesting.oaipmh.CloseableOaiClient;
+import eu.europeana.metis.harvesting.oaipmh.OaiHarvest;
+import eu.europeana.metis.harvesting.oaipmh.OaiHarvester;
+import eu.europeana.metis.harvesting.oaipmh.OaiHarvesterImpl;
+import eu.europeana.metis.harvesting.oaipmh.OaiHarvesterImpl.ConnectionClientFactory;
+import eu.europeana.metis.harvesting.oaipmh.OaiRecordHeaderIterator;
+import eu.europeana.metis.harvesting.oaipmh.OaiRepository;
+import eu.europeana.metis.sandbox.common.exception.ServiceException;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+
+@Service
+public class HarvestServiceOaiPmhImpl implements HarvestServiceOaiPmh {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HarvestServiceOaiPmhImpl.class);
+
+  private static ConnectionClientFactory CONNECTION_CLIENT_FACTORY;
+
+  public List<ByteArrayInputStream> harvestOaiPmh(String endpoint, String setSpec, String prefix) {
+
+    CloseableOaiClient client = CONNECTION_CLIENT_FACTORY.createConnectionClient(endpoint);
+    OaiHarvester oaiHarvester = new OaiHarvesterImpl(CONNECTION_CLIENT_FACTORY);
+    List<ByteArrayInputStream> records = new ArrayList<>();
+    List<Pair<String, Exception>> exception = new ArrayList<>(1);
+
+    try {
+      OaiRecordHeaderIterator recordHeaderIterator =
+          oaiHarvester.harvestRecordHeaders(new OaiHarvest("repoURL", "prefix", "spec "));
+
+      recordHeaderIterator.forEach(r -> {
+        OaiRepository oaiRepo = new OaiRepository("url", "prefix");
+        try {
+          var rec = oaiHarvester.harvestRecord(oaiRepo, r.getOaiIdentifier());
+          records.add(new ByteArrayInputStream(rec.getRecord().readAllBytes()));
+        } catch (HarvesterException | IOException e) {
+          exception.add(new ImmutablePair<>(r.getOaiIdentifier(), e));
+          return ReportingIteration.IterationResult.TERMINATE;
+        }
+        return ReportingIteration.IterationResult.CONTINUE;
+      });
+      if (!exception.isEmpty()) {
+        throw new ServiceException("Error harvesting records " + exception.get(0).getKey(),
```

Comment:
Is it possible to be more specific with the error messages? The one below is very similar and the more specific it is, the better for debugging

Reply:
if there was an exception iterating OAIrepo throw exception with specific key and value of the record

Reply:
Done